rotkehlchen/chain/gnosis/modules/sdai/decoder.py

```diff
@@ -0,0 +1,128 @@
+from rotkehlchen.constants.assets import A_SDAI
+from rotkehlchen.types import ChainID
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface, ReloadableCacheDecoderMixin
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.types import ChecksumEvmAddress
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.gnosis.node_inquirer import GnosisInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+SDAI_DEPOSIT = b'\xdc\xbc\x1c\x05$\x0f1\xff:\xd0g\xef\x1e\xe3\\\xe4\x99wbu.:\tR\x84uED\xf4\xc7\t\xd7'
+SDAI_REDEEM = b'\xfb\xdey} \x1ch\x1b\x91\x05e)\x11\x9e\x0b\x02@|{\xb9jJ,u\xc0\x1f\xc9fr2\xc8\xdb'
+
+
+class SdaiDecoder(DecoderInterface, ReloadableCacheDecoderMixin):
+    def __init__(
+            self,
+            gnosis_inquirer: 'GnosisInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=gnosis_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.sdai = A_SDAI.resolve_to_evm_token()
+
+    def decode_sdai_events(self, context: DecoderContext) -> DecodingOutput:
+        if context.tx_log.topics[0] == SDAI_DEPOSIT:
+            from_address = context.tx_log.topics[1]
+            to_address = context.tx_log.topics[2]
+            owner_address = to_address
+        elif context.tx_log.topics[0] == SDAI_REDEEM:
+            from_address = None
+            to_address = None
+            owner_address = None
+        else:
+            return DEFAULT_DECODING_OUTPUT
+
+        if from_address is None or to_address is None:
+            return DEFAULT_DECODING_OUTPUT
+
+        amount = asset_normalized_value(
+            amount=hex_or_bytes_to_int(context.tx_log.data[32:64]),
+            asset=self.sdai,
+        )
+        if amount == ZERO:
+            return DEFAULT_DECODING_OUTPUT
+
+        if owner_address == from_address:
+            event_type = HistoryEventType.DEPOSIT
+            event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+            notes = f'Return {amount} sDAI to sDAI contract'
+        else:
+            event_type = HistoryEventType.WITHDRAWAL
+            event_subtype = HistoryEventSubType.REMOVE_ASSET
+            notes = f'Withdraw {amount} sDAI from sDAI contract'
+
+        transfer = self.base.make_event_from_transaction(
+            transaction=context.transaction,
+            tx_log=context.tx_log,
+            event_type=event_type,
+            event_subtype=event_subtype,
+            asset=self.sdai,
+            balance=Balance(amount=amount),
+            location_label=owner_address,
+            notes=notes,
+            address=self.sdai.evm_address,
+            counterparty=CPT_SDAI,
```

Comment:
This is not needed in your case. Creating an event is required when the contract does something that we don't decode because doesn't affect the user address, so basically when you don't have a spend/receive . If you check decoded events at the end of the transcation you have 3 events

```
{'identifier': None, 'entry_type': 'evm event', 'event_identifier': '1000x1342646cab122d58f0b7dfae404dad5235d42224de881099dc05e59477bb93aa', 'sequence_index': 0, 'timestamp': 1707169525000, 'location': 'gnosis', 'asset': 'XDAI', 'balance': {'amount': '0.000367251244452481', 'usd_value': '0'}, 'event_type': 'spend', 'event_subtype': 'fee', 'location_label': '0x78E87757861185Ec5e8C0EF6BF0C69Fa7832df6C', 'notes': 'Burned 0.000367251244452481 XDAI for gas', 'tx_hash': '0x1342646cab122d58f0b7dfae404dad5235d42224de881099dc05e59477bb93aa', 'counterparty': 'gas', 'product': None, 'address': None, 'extra_data': None}
{'identifier': None, 'entry_type': 'evm event', 'event_identifier': '1000x1342646cab122d58f0b7dfae404dad5235d42224de881099dc05e59477bb93aa', 'sequence_index': 1, 'timestamp': 1707169525000, 'location': 'gnosis', 'asset': 'XDAI', 'balance': {'amount': '315', 'usd_value': '0'}, 'event_type': 'spend', 'event_subtype': 'none', 'location_label': '0x78E87757861185Ec5e8C0EF6BF0C69Fa7832df6C', 'notes': 'Send 315 XDAI to 0xD499b51fcFc66bd31248ef4b28d656d67E591A94', 'tx_hash': '0x1342646cab122d58f0b7dfae404dad5235d42224de881099dc05e59477bb93aa', 'counterparty': None, 'product': None, 'address': '0xD499b51fcFc66bd31248ef4b28d656d67E591A94', 'extra_data': None}
{'identifier': None, 'entry_type': 'evm event', 'event_identifier': '1000x1342646cab122d58f0b7dfae404dad5235d42224de881099dc05e59477bb93aa', 'sequence_index': 3020, 'timestamp': 1707169525000, 'location': 'gnosis', 'asset': 'eip155:100/erc20:0xaf204776c7245bF4147c2612BF6e5972Ee483701', 'balance': {'amount': '303.052244055946806232', 'usd_value': '0'}, 'event_type': 'receive', 'event_subtype': 'none', 'location_label': '0x78E87757861185Ec5e8C0EF6BF0C69Fa7832df6C', 'notes': 'Receive 303.052244055946806232 sDAI from 0x0000000000000000000000000000000000000000 to 0x78E87757861185Ec5e8C0EF6BF0C69Fa7832df6C', 'tx_hash': '0x1342646cab122d58f0b7dfae404dad5235d42224de881099dc05e59477bb93aa', 'counterparty': None, 'product': None, 'address': '0x0000000000000000000000000000000000000000', 'extra_data': None}
```

that are exactly what you need